common/lib/dependabot/new_version.rb

```diff
@@ -0,0 +1,228 @@
+
+    abstract!
+
+    VersionParameter = T.type_alias { T.nilable(T.any(String, Integer, Gem::Version)) }
+
+    sig { override.overridable.params(version: VersionParameter, version_string: T.nilable(String)).void }
+    def initialize(version, version_string = nil)
+      @version_string = T.let(version_string || version.to_s, String)
+      super(version)
+    end
+
+    sig { override.overridable.params(version: VersionParameter).returns(Dependabot::Version) }
+    def self.new(version)
+      T.cast(super, Dependabot::NewVersion)
+      super
+    end
+
+    # Opt-in to Rubygems 4 behavior
+    sig { override.overridable.params(version: VersionParameter).returns(T::Boolean) }
+    def self.correct?(version)
+      return false if version.nil?
+
+      version.to_s.match?(ANCHORED_VERSION_PATTERN)
+    end
+
+    sig { override.overridable.returns(String) }
+    def to_semver
+      @version_string
+    end
+
+    sig { override.overridable.returns(String) }
+    def inspect
+      "#<#{self.class} #{@version_string}>"
+    end
+
+    sig { override.overridable.returns(String) }
+    def to_s
+      @version_string
+    end
+
+    sig { override.overridable.returns(T::Boolean) }
+    def prerelease?
+      tokens.any? do |token|
+        next true if token == "eap"
+        next false unless NAMED_QUALIFIERS_HIERARCHY[T.must(token)]
+
+        T.must(NAMED_QUALIFIERS_HIERARCHY[T.must(token)]) < 6
+      end
+    end
+
+    # rubocop:disable  Metrics/AbcSize
+    sig { params(other: T.any(String, Integer, Gem::Version)).returns(Integer) }
+    def <=>(other)
+      version = stringify_version(@version_string)
+      version = fill_tokens(version)
+      version = trim_version(version)
+
+      other_version = stringify_version(other)
+      other_version = fill_tokens(other_version)
+      other_version = trim_version(other_version)
+
+      version, other_version = convert_dates(version, other_version)
+
+      prefixed_tokens = split_into_prefixed_tokens(T.must(version))
+      other_prefixed_tokens = split_into_prefixed_tokens(T.must(other_version))
+
+      prefixed_tokens, other_prefixed_tokens =
+        pad_for_comparison(prefixed_tokens, other_prefixed_tokens)
+
+      T.must(prefixed_tokens).count.times.each do |index|
+        prefix = T.must(T.must(prefixed_tokens)[index])[0]
+        token = T.must(T.must(prefixed_tokens)[index])[1..-1] || ""
+        other_prefix = T.must(T.must(other_prefixed_tokens)[index])[0]
+        other_token = T.must(T.must(other_prefixed_tokens)[index])[1..-1] || ""
+        comp = compare_prefixed_token(
+          prefix: prefix.to_s,
+          token: token,
+          other_prefix: other_prefix.to_s,
+          other_token: other_token
+        )
+        return T.must(comp) unless T.must(comp).zero?
+      end
+
+      0
+    end
+    # rubocop:enable  Metrics/AbcSize
+
+    private
+
+    sig { returns(T::Array[T.nilable(String)]) }
+    def tokens
+      version = @version_string.to_s.downcase
```

Comment:
the @version_string param is used in a few methods in new_version. For example here. @abdulapopoola

Reply:
How does it actually differ from the version? I'd like us to understand how and why it is used as we standardize on this.

Reply:
Discussed offline but to close on this @version_string is the raw version that we get when we fetch the dependency from a repo. In some cases it's not a valid Gem::Version though it still carries important info for deciding if a version is a prerelease, for example. @version_string could be "2.10.0.pr3" and version (Gem::Version) would be "2.10.0"